File: samples/language-impl-comparison/ocaml/yaml_parser.ml
(* YAML風パーサー：インデント管理が重要な題材。

   対応する構文（簡易版）：
   - スカラー値: 文字列、数値、真偽値、null
   - リスト: `- item1`
   - マップ: `key: value`
   - ネストしたインデント構造

   インデント処理の特徴：
   - OCamlの関数型スタイルとパターンマッチを活用
   - エラー回復機能でインデントミスを報告しつつ継続 *)

module StringMap = Map.Make(String)

(* YAML値の表現。 *)
type yaml_value =
  | Scalar of string
  | List of yaml_value list
  | Map of yaml_value StringMap.t
  | Null

type document = yaml_value

(* パーサー型。 *)
type 'a parser = string -> int -> ('a * int) option

exception ParseError of string

(* パーサーヘルパー関数 *)

let peek input pos =
  if pos < String.length input then
    Some input.[pos]
  else
    None

let advance pos = pos + 1

let is_eof input pos = pos >= String.length input

let expect input pos expected =
  match peek input pos with
  | Some c when c = expected -> advance pos
  | _ -> raise (ParseError ("期待された文字 '" ^ String.make 1 expected ^ "' が見つかりません"))

let expect_string input pos expected =
  let len = String.length expected in
  if pos + len <= String.length input && String.sub input pos len = expected then
    pos + len
  else
    raise (ParseError ("期待された文字列 '" ^ expected ^ "' が見つかりません"))

(* 水平空白のみをスキップ（改行は含まない）。 *)
let rec hspace input pos =
  match peek input pos with
  | Some (' ' | '\t') -> hspace input (advance pos)
  | _ -> pos

(* 改行をスキップ。 *)
let newline input pos =
  match peek input pos with
  | Some '\n' -> advance pos
  | Some '\r' ->
    let pos' = advance pos in
    (match peek input pos' with
    | Some '\n' -> advance pos'
    | _ -> pos')
  | _ -> pos

(* コメントのスキップ（`#` から行末まで）。 *)
let comment input pos =
  match peek input pos with
  | Some '#' ->
    let rec skip pos =
      match peek input pos with
      | Some '\n' | None -> pos
      | _ -> skip (advance pos)
    in skip (advance pos)
  | _ -> pos

(* 空行またはコメント行をスキップ。 *)
let blank_or_comment input pos =
  let pos = hspace input pos in
  let pos = comment input pos in
  newline input pos

(* 特定のインデントレベルを期待する。 *)
let expect_indent input pos level =
  let rec count_spaces pos n =
    match peek input pos with
    | Some ' ' -> count_spaces (advance pos) (n + 1)
    | _ -> (pos, n)
  in
  let (pos', spaces) = count_spaces pos 0 in
  if spaces = level then
    pos'
  else
    raise (ParseError ("インデント不一致: 期待 " ^ string_of_int level ^ ", 実際 " ^ string_of_int spaces))

(* 現在よりも深いインデントを検出。 *)
let deeper_indent input pos current =
  let rec count_spaces pos n =
    match peek input pos with
    | Some ' ' -> count_spaces (advance pos) (n + 1)
    | _ -> (pos, n)
  in
  let (pos', spaces) = count_spaces pos 0 in
  if spaces > current then
    (pos', spaces)
  else
    raise (ParseError ("深いインデントが期待されます: 現在 " ^ string_of_int current ^ ", 実際 " ^ string_of_int spaces))

(* スカラー値のパース（前方宣言）。 *)
let scalar_value input pos = None

(* YAML値のパース（前方宣言）。 *)
let rec parse_value input pos indent = None

(* スカラー値のパース実装。 *)
let scalar_value input pos =
  (* null *)
  if pos + 4 <= String.length input && String.sub input pos 4 = "null" then
    Some (Null, pos + 4)
  else if pos < String.length input && input.[pos] = '~' then
    Some (Null, advance pos)
  (* 真偽値 *)
  else if pos + 4 <= String.length input && String.sub input pos 4 = "true" then
    Some (Scalar "true", pos + 4)
  else if pos + 5 <= String.length input && String.sub input pos 5 = "false" then
    Some (Scalar "false", pos + 5)
  (* 数値（簡易実装） *)
  else
    let rec parse_number pos acc =
      match peek input pos with
      | Some ('0'..'9' as c) -> parse_number (advance pos) (acc ^ String.make 1 c)
      | _ -> if acc = "" then None else Some (Scalar acc, pos)
    in
    match parse_number pos "" with
    | Some result -> Some result
    | None ->
      (* 文字列（引用符付き） *)
      if pos < String.length input && input.[pos] = '"' then
        let rec parse_quoted pos acc =
          match peek input pos with
          | Some '"' -> Some (Scalar acc, advance pos)
          | Some c -> parse_quoted (advance pos) (acc ^ String.make 1 c)
          | None -> raise (ParseError "引用符が閉じられていません")
        in
        parse_quoted (advance pos) ""
      else
        (* 文字列（引用符なし：行末または `:` まで） *)
        let rec parse_unquoted pos acc =
          match peek input pos with
          | Some ('\n' | ':' | '#') -> Some (Scalar (String.trim acc), pos)
          | Some c -> parse_unquoted (advance pos) (acc ^ String.make 1 c)
          | None -> Some (Scalar (String.trim acc), pos)
        in
        parse_unquoted pos ""

(* リスト項目のパース（`- value` 形式）。 *)
and parse_list_item input pos indent =
  let pos = expect_indent input pos indent in
  let pos = expect input pos '-' in
  let pos = hspace input pos in
  parse_value input pos (indent + 2)

(* リスト全体のパース。 *)
and parse_list input pos indent =
  let rec parse_items pos acc =
    try
      match parse_list_item input pos indent with
      | Some (item, pos') ->
        let pos'' = (match peek input pos' with Some '\n' -> newline input pos' | _ -> pos') in
        parse_items pos'' (item :: acc)
      | None -> (List.rev acc, pos)
    with ParseError _ -> (List.rev acc, pos)
  in
  let (items, pos') = parse_items pos [] in
  if items = [] then
    None
  else
    Some (List items, pos')

(* マップのキーバリューペアのパース（`key: value` 形式）。 *)
and parse_map_entry input pos indent =
  let pos = expect_indent input pos indent in
  let rec parse_key pos acc =
    match peek input pos with
    | Some (':' | '\n') -> (String.trim acc, pos)
    | Some c -> parse_key (advance pos) (acc ^ String.make 1 c)
    | None -> (String.trim acc, pos)
  in
  let (key, pos) = parse_key pos "" in
  let pos = expect input pos ':' in
  let pos = hspace input pos in
  (* 同じ行に値があるか、次の行にネストされているか *)
  let (value, pos') =
    match peek input pos with
    | Some '\n' ->
      let pos' = newline input pos in
      (match parse_value input pos' (indent + 2) with
      | Some (v, p) -> (v, p)
      | None -> raise (ParseError "値が期待されます"))
    | _ ->
      (match parse_value input pos indent with
      | Some (v, p) -> (v, p)
      | None -> raise (ParseError "値が期待されます"))
  in
  Some ((key, value), pos')

(* マップ全体のパース。 *)
and parse_map input pos indent =
  let rec parse_entries pos acc =
    try
      match parse_map_entry input pos indent with
      | Some ((key, value), pos') ->
        let pos'' = (match peek input pos' with Some '\n' -> newline input pos' | _ -> pos') in
        parse_entries pos'' ((key, value) :: acc)
      | None -> (List.rev acc, pos)
    with ParseError _ -> (List.rev acc, pos)
  in
  let (entries, pos') = parse_entries pos [] in
  if entries = [] then
    None
  else
    let map = List.fold_left (fun m (k, v) -> StringMap.add k v m) StringMap.empty entries in
    Some (Map map, pos')

(* YAML値のパース（再帰的）実装。 *)
and parse_value input pos indent =
  match parse_list input pos indent with
  | Some result -> Some result
  | None ->
    match parse_map input pos indent with
    | Some result -> Some result
    | None -> scalar_value input pos

(* ドキュメント全体のパース。 *)
let document input =
  let rec skip_blanks pos =
    if is_eof input pos then pos
    else
      try
        let pos' = blank_or_comment input pos in
        if pos' = pos then pos else skip_blanks pos'
      with ParseError _ -> pos
  in
  let pos = skip_blanks 0 in
  match parse_value input pos 0 with
  | Some (doc, pos') ->
    let pos'' = skip_blanks pos' in
    if is_eof input pos'' then
      Some doc
    else
      raise (ParseError "ドキュメントの終端が期待されます")
  | None -> None

(* パブリックAPI：YAML文字列をパース。 *)
let parse_yaml input =
  try
    document input
  with ParseError msg ->
    print_endline ("パースエラー: " ^ msg);
    None

(* 簡易的なレンダリング（検証用）。 *)
let render_to_string doc =
  let rec render_value value indent =
    let indent_str = String.make indent ' ' in
    match value with
    | Scalar s -> s
    | Null -> "null"
    | List items ->
      String.concat "\n" (List.map (fun item -> indent_str ^ "- " ^ render_value item (indent + 2)) items)
    | Map entries ->
      let entries_list = StringMap.bindings entries in
      String.concat "\n" (List.map (fun (key, val_) ->
        match val_ with
        | Scalar _ | Null -> indent_str ^ key ^ ": " ^ render_value val_ 0
        | _ -> indent_str ^ key ^ ":\n" ^ render_value val_ (indent + 2)
      ) entries_list)
  in
  render_value doc 0

(* テスト例。 *)
let test_examples () =
  let examples = [
    ("simple_scalar", "hello");
    ("simple_list", "- item1\n- item2\n- item3");
    ("simple_map", "key1: value1\nkey2: value2");
    ("nested_map", "parent:\n  child1: value1\n  child2: value2");
    ("nested_list", "items:\n  - item1\n  - item2");
    ("mixed", "name: John\nage: 30\nhobbies:\n  - reading\n  - coding")
  ] in
  List.iter (fun (name, yaml_str) ->
    print_endline ("--- " ^ name ^ " ---");
    match parse_yaml yaml_str with
    | Some doc ->
      print_endline "パース成功:";
      print_endline (render_to_string doc)
    | None ->
      print_endline "パースエラー"
  ) examples

(* インデント処理の課題と解決策：

   1. **インデントレベルの追跡**
      - パーサー引数としてインデントレベルを渡す
      - OCamlの純粋関数型スタイルでパーサー状態を管理

   2. **エラー回復**
      - optionとexceptionでバックトラックを制御
      - ParseError例外で分かりやすいエラーメッセージを提供

   3. **空白の扱い**
      - hspaceで水平空白のみをスキップ（改行は構文の一部）
      - newlineでCR/LF/CRLFを正規化

   Remlとの比較：

   - **OCamlの利点**:
     - 強力なパターンマッチング
     - 効率的なコンパイル済みコード

   - **OCamlの課題**:
     - パーサーコンビネーターライブラリがRemlほど充実していない
     - 手動のバックトラック管理が煩雑

   - **Remlの利点**:
     - 字句レイヤの柔軟性により、インデント処理が自然に表現できる
     - cut/commitによるエラー品質の向上
     - recoverによる部分的なパース継続が可能 *)

File: samples/language-impl-comparison/rust/yaml_parser.rs
use std::collections::HashMap;

/// YAML風パーサー：インデント管理が重要な題材。
///
/// 対応する構文（簡易版）：
/// - スカラー値: 文字列、数値、真偽値、null
/// - リスト: `- item1`
/// - マップ: `key: value`
/// - ネストしたインデント構造
///
/// インデント処理の特徴：
/// - Rustの所有権システムとResultを活用したパーサー実装
/// - エラー回復機能でインデントミスを報告しつつ継続

#[derive(Debug, Clone, PartialEq)]
pub enum YamlValue {
    Scalar(String),
    List(Vec<YamlValue>),
    Map(HashMap<String, YamlValue>),
    Null,
}

pub type Document = YamlValue;

#[derive(Debug, Clone)]
pub struct ParseError {
    message: String,
}

impl ParseError {
    fn new(message: impl Into<String>) -> Self {
        ParseError {
            message: message.into(),
        }
    }
}

impl std::fmt::Display for ParseError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{}", self.message)
    }
}

impl std::error::Error for ParseError {}

pub struct Parser<'a> {
    input: &'a str,
    pos: usize,
}

impl<'a> Parser<'a> {
    pub fn new(input: &'a str) -> Self {
        Parser { input, pos: 0 }
    }

    fn peek(&self) -> Option<char> {
        self.input.chars().nth(self.pos)
    }

    fn advance(&mut self) {
        if self.pos < self.input.len() {
            self.pos += self.input.chars().nth(self.pos).unwrap().len_utf8();
        }
    }

    fn is_eof(&self) -> bool {
        self.pos >= self.input.len()
    }

    fn expect(&mut self, expected: char) -> Result<(), ParseError> {
        if self.peek() != Some(expected) {
            return Err(ParseError::new(format!(
                "期待された文字 '{}' が見つかりません",
                expected
            )));
        }
        self.advance();
        Ok(())
    }

    fn expect_string(&mut self, expected: &str) -> Result<(), ParseError> {
        for c in expected.chars() {
            self.expect(c)?;
        }
        Ok(())
    }

    /// 水平空白のみをスキップ（改行は含まない）。
    fn hspace(&mut self) {
        while let Some(c) = self.peek() {
            if c == ' ' || c == '\t' {
                self.advance();
            } else {
                break;
            }
        }
    }

    /// 改行をスキップ。
    fn newline(&mut self) {
        if self.peek() == Some('\n') {
            self.advance();
        } else if self.peek() == Some('\r') {
            self.advance();
            if self.peek() == Some('\n') {
                self.advance();
            }
        }
    }

    /// コメントのスキップ（`#` から行末まで）。
    fn comment(&mut self) {
        if self.peek() == Some('#') {
            self.advance();
            while let Some(c) = self.peek() {
                if c == '\n' {
                    break;
                }
                self.advance();
            }
        }
    }

    /// 空行またはコメント行をスキップ。
    fn blank_or_comment(&mut self) {
        self.hspace();
        self.comment();
        self.newline();
    }

    /// 特定のインデントレベルを期待する。
    fn expect_indent(&mut self, level: usize) -> Result<(), ParseError> {
        let mut spaces = 0;
        while self.peek() == Some(' ') {
            spaces += 1;
            self.advance();
        }

        if spaces != level {
            return Err(ParseError::new(format!(
                "インデント不一致: 期待 {}, 実際 {}",
                level, spaces
            )));
        }

        Ok(())
    }

    /// 現在よりも深いインデントを検出。
    fn deeper_indent(&mut self, current: usize) -> Result<usize, ParseError> {
        let mut spaces = 0;
        while self.peek() == Some(' ') {
            spaces += 1;
            self.advance();
        }

        if spaces <= current {
            return Err(ParseError::new(format!(
                "深いインデントが期待されます: 現在 {}, 実際 {}",
                current, spaces
            )));
        }

        Ok(spaces)
    }

    /// スカラー値のパース。
    fn scalar_value(&mut self) -> Result<YamlValue, ParseError> {
        // null
        if self.input[self.pos..].starts_with("null") {
            self.expect_string("null")?;
            return Ok(YamlValue::Null);
        }

        if self.peek() == Some('~') {
            self.advance();
            return Ok(YamlValue::Null);
        }

        // 真偽値
        if self.input[self.pos..].starts_with("true") {
            self.expect_string("true")?;
            return Ok(YamlValue::Scalar("true".to_string()));
        }

        if self.input[self.pos..].starts_with("false") {
            self.expect_string("false")?;
            return Ok(YamlValue::Scalar("false".to_string()));
        }

        // 数値（簡易実装）
        let mut num_str = String::new();
        while let Some(c) = self.peek() {
            if c.is_ascii_digit() {
                num_str.push(c);
                self.advance();
            } else {
                break;
            }
        }

        if !num_str.is_empty() {
            return Ok(YamlValue::Scalar(num_str));
        }

        // 文字列（引用符付き）
        if self.peek() == Some('"') {
            self.advance();
            let mut str = String::new();
            while let Some(c) = self.peek() {
                if c == '"' {
                    break;
                }
                str.push(c);
                self.advance();
            }
            self.expect('"')?;
            return Ok(YamlValue::Scalar(str));
        }

        // 文字列（引用符なし：行末または `:` まで）
        let mut str = String::new();
        while let Some(c) = self.peek() {
            if c == '\n' || c == ':' || c == '#' {
                break;
            }
            str.push(c);
            self.advance();
        }

        Ok(YamlValue::Scalar(str.trim().to_string()))
    }

    /// リスト項目のパース（`- value` 形式）。
    fn parse_list_item(&mut self, indent: usize) -> Result<YamlValue, ParseError> {
        self.expect_indent(indent)?;
        self.expect('-')?;
        self.hspace();
        self.parse_value(indent + 2)
    }

    /// リスト全体のパース。
    fn parse_list(&mut self, indent: usize) -> Result<YamlValue, ParseError> {
        let mut items = Vec::new();

        loop {
            let saved_pos = self.pos;
            match self.parse_list_item(indent) {
                Ok(item) => {
                    items.push(item);
                    if self.peek() == Some('\n') {
                        self.newline();
                    } else {
                        break;
                    }
                }
                Err(_) => {
                    self.pos = saved_pos;
                    break;
                }
            }
        }

        if items.is_empty() {
            return Err(ParseError::new("リストが空です"));
        }

        Ok(YamlValue::List(items))
    }

    /// マップのキーバリューペアのパース（`key: value` 形式）。
    fn parse_map_entry(&mut self, indent: usize) -> Result<(String, YamlValue), ParseError> {
        self.expect_indent(indent)?;

        let mut key = String::new();
        while let Some(c) = self.peek() {
            if c == ':' || c == '\n' {
                break;
            }
            key.push(c);
            self.advance();
        }

        let key = key.trim().to_string();
        self.expect(':')?;
        self.hspace();

        // 同じ行に値があるか、次の行にネストされているか
        let value = if self.peek() == Some('\n') {
            self.newline();
            self.parse_value(indent + 2)?
        } else {
            self.parse_value(indent)?
        };

        Ok((key, value))
    }

    /// マップ全体のパース。
    fn parse_map(&mut self, indent: usize) -> Result<YamlValue, ParseError> {
        let mut entries = HashMap::new();

        loop {
            let saved_pos = self.pos;
            match self.parse_map_entry(indent) {
                Ok((key, value)) => {
                    entries.insert(key, value);
                    if self.peek() == Some('\n') {
                        self.newline();
                    } else {
                        break;
                    }
                }
                Err(_) => {
                    self.pos = saved_pos;
                    break;
                }
            }
        }

        if entries.is_empty() {
            return Err(ParseError::new("マップが空です"));
        }

        Ok(YamlValue::Map(entries))
    }

    /// YAML値のパース（再帰的）。
    fn parse_value(&mut self, indent: usize) -> Result<YamlValue, ParseError> {
        let saved_pos = self.pos;

        // リストを試行
        if let Ok(list) = self.parse_list(indent) {
            return Ok(list);
        }

        self.pos = saved_pos;

        // マップを試行
        if let Ok(map) = self.parse_map(indent) {
            return Ok(map);
        }

        self.pos = saved_pos;

        // スカラー
        self.scalar_value()
    }

    /// ドキュメント全体のパース。
    fn document(&mut self) -> Result<Document, ParseError> {
        // 空行やコメントをスキップ
        while !self.is_eof() {
            let saved_pos = self.pos;
            self.blank_or_comment();
            if self.pos == saved_pos {
                break;
            }
        }

        let doc = self.parse_value(0)?;

        // 末尾の空行やコメントをスキップ
        while !self.is_eof() {
            let saved_pos = self.pos;
            self.blank_or_comment();
            if self.pos == saved_pos {
                break;
            }
        }

        if !self.is_eof() {
            return Err(ParseError::new("ドキュメントの終端が期待されます"));
        }

        Ok(doc)
    }
}

/// パブリックAPI：YAML文字列をパース。
pub fn parse_yaml(input: &str) -> Result<Document, ParseError> {
    let mut parser = Parser::new(input);
    parser.document()
}

/// 簡易的なレンダリング（検証用）。
pub fn render_to_string(doc: &Document) -> String {
    fn render_value(value: &YamlValue, indent: usize) -> String {
        let indent_str = " ".repeat(indent);

        match value {
            YamlValue::Scalar(s) => s.clone(),
            YamlValue::Null => "null".to_string(),
            YamlValue::List(items) => items
                .iter()
                .map(|item| format!("{}- {}", indent_str, render_value(item, indent + 2)))
                .collect::<Vec<_>>()
                .join("\n"),
            YamlValue::Map(entries) => {
                let mut lines = Vec::new();
                for (key, val) in entries.iter() {
                    match val {
                        YamlValue::Scalar(_) | YamlValue::Null => {
                            lines.push(format!("{}{}: {}", indent_str, key, render_value(val, 0)));
                        }
                        _ => {
                            lines.push(format!(
                                "{}{}:\n{}",
                                indent_str,
                                key,
                                render_value(val, indent + 2)
                            ));
                        }
                    }
                }
                lines.join("\n")
            }
        }
    }

    render_value(doc, 0)
}

/// テスト例。
pub fn test_examples() {
    let examples = vec![
        ("simple_scalar", "hello"),
        ("simple_list", "- item1\n- item2\n- item3"),
        ("simple_map", "key1: value1\nkey2: value2"),
        ("nested_map", "parent:\n  child1: value1\n  child2: value2"),
        ("nested_list", "items:\n  - item1\n  - item2"),
        ("mixed", "name: John\nage: 30\nhobbies:\n  - reading\n  - coding"),
    ];

    for (name, yaml_str) in examples {
        println!("--- {} ---", name);
        match parse_yaml(yaml_str) {
            Ok(doc) => {
                println!("パース成功:");
                println!("{}", render_to_string(&doc));
            }
            Err(err) => {
                println!("パースエラー: {}", err);
            }
        }
    }
}

/// インデント処理の課題と解決策：
///
/// 1. **インデントレベルの追跡**
///    - パーサー引数としてインデントレベルを渡す
///    - Rustの所有権システムでパーサー状態を管理
///
/// 2. **エラー回復**
///    - Resultでバックトラックを制御
///    - ParseError型で分かりやすいエラーメッセージを提供
///
/// 3. **空白の扱い**
///    - hspaceで水平空白のみをスキップ（改行は構文の一部）
///    - newlineでCR/LF/CRLFを正規化
///
/// Remlとの比較：
///
/// - **Rustの利点**:
///   - ゼロコスト抽象化と高速な実行
///   - 強力な所有権システムによるメモリ安全性
///
/// - **Rustの課題**:
///   - パーサーコンビネーターライブラリがRemlほど充実していない
///   - 手動のバックトラック管理が煩雑
///
/// - **Remlの利点**:
///   - 字句レイヤの柔軟性により、インデント処理が自然に表現できる
///   - cut/commitによるエラー品質の向上
///   - recoverによる部分的なパース継続が可能